File: problems/search/PROB-SEARCH-001/tests/python/test_membership.py
from solution import linear_contains, binary_contains, hash_contains


def check_all(values, target, expected):
    assert linear_contains(values, target) is expected
    assert binary_contains(sorted(values), target) is expected
    assert hash_contains(set(values), target) is expected


def test_nominal_present():
    check_all([1, 3, 5, 7, 9], 7, True)


def test_nominal_absent():
    check_all([1, 3, 5, 7, 9], 4, False)


def test_empty():
    check_all([], 1, False)


def test_single_present():
    check_all([42], 42, True)


def test_duplicates():
    check_all([2, 2, 2, 3], 2, True)


def test_negative_values():
    check_all([-5, -1, 0, 4], -1, True)

File: problems/search/PROB-SEARCH-001/tests/cpp/test_membership.cpp
#include <cassert>
#include <unordered_set>
#include <vector>

bool linear_contains(const std::vector<int>& values, int target);
bool binary_contains(const std::vector<int>& sorted_values, int target);
bool hash_contains(const std::unordered_set<int>& values, int target);

int main() {
    const std::vector<int> values{1, 3, 5, 7, 9};
    const std::unordered_set<int> set{1, 3, 5, 7, 9};

    assert(linear_contains(values, 7));
    assert(binary_contains(values, 7));
    assert(hash_contains(set, 7));

    assert(!linear_contains(values, 4));
    assert(!binary_contains(values, 4));
    assert(!hash_contains(set, 4));

    const std::vector<int> empty;
    const std::unordered_set<int> empty_set;
    assert(!linear_contains(empty, 1));
    assert(!binary_contains(empty, 1));
    assert(!hash_contains(empty_set, 1));

    const std::vector<int> one{42};
    const std::unordered_set<int> one_set{42};
    assert(linear_contains(one, 42));
    assert(binary_contains(one, 42));
    assert(hash_contains(one_set, 42));

    return 0;
}
